pkg/resource/stack/secrets.go

```diff
@@ -86,171 +89,507 @@ func (s NamedStackSecretsProvider) OfType(ty string, state json.RawMessage) (sec
+//
+//	cache := NewSecretCache()
+//	secret := &resource.Secret{}
+//	cache.Write("plaintext", "ciphertext", secret)
+//	plaintext, ok := cache.TryDecrypt("ciphertext") // "plaintext", true
+//	ciphertext, ok := cache.TryEncrypt(secret, "plaintext") // "ciphertext", true
+func NewSecretCache() SecretCache {
+	// If the environment variable PULUMI_DISABLE_SECRET_CACHE is set to "true", the secret cache will be disabled and
+	// no entries will be stored. This is a short-term escape hatch in case there's unforeseen issues with expanded
+	// caching scopes and should be removable once we're confident that customers are not affected.
+	if os.Getenv("PULUMI_DISABLE_SECRET_CACHE") == "true" {
+		return nullSecretCache{}
 	}
-	ciphertext, err := csm.manager.Encrypter().EncryptValue(ctx, plaintext)
-	if err != nil {
-		return "", err
+	return &secretCache{}
+}
+
+// Write stores the plaintext, ciphertext, and secret in the cache, overwriting any previous entry for the secret.
+// This method is thread-safe and can be called concurrently by multiple goroutines.
+func (c *secretCache) Write(plaintext, ciphertext string, secret *resource.Secret) {
+	entry := secretCacheEntry{plaintext, ciphertext, secret}
+	c.bySecret.Store(secret, entry)
+	c.byCiphertext.Store(ciphertext, entry)
+}
+
+// LookupCiphertext returns the cached ciphertext for the given secret and plaintext, if it exists.
+// The ciphertext is returned as a string, and a boolean is returned to indicate whether the secret was found.
+// This method is thread-safe and can be called concurrently by multiple goroutines.
+func (c *secretCache) LookupCiphertext(secret *resource.Secret, plaintext string) (string, bool) {
+	entry, ok := c.bySecret.Load(secret)
+	if !ok {
+		return "", false
+	}
+	cacheEntry := entry.(secretCacheEntry)
+	if cacheEntry.plaintext != plaintext {
+		return "", false
+	}
+	return cacheEntry.ciphertext, true
+}
+
+// LookupPlaintext returns the cached plaintext for the given ciphertext, if it exists.
+// The plaintext is returned as a string, and a boolean is returned to indicate whether the ciphertext was found.
+// This method is thread-safe and can be called concurrently by multiple goroutines.
+func (c *secretCache) LookupPlaintext(ciphertext string) (string, bool) {
+	entry, ok := c.byCiphertext.Load(ciphertext)
+	if !ok {
+		return "", false
 	}
-	csm.insert(secret, plaintext, ciphertext)
-	return ciphertext, nil
+	return entry.(secretCacheEntry).plaintext, true
+}
+
+// BatchEncrypter is a wrapper around an Encrypter which will then call the BatchEncrypt method on the underlying
+// Encrypter. This allows encryption operations to be batched to avoid round-trips for when the underlying encrypter
+// requires a network call. Encryptions to a secret are enqueued and processed in a batch request either when the queue
+// is full or when the wrapping transaction is committed.
```

Comment:
I feel like this comment is more about an implementation than the interface. A `BatchEncrypter` is just an `Encrypter` that supports enqueuing encryptions for batch processing -- how that happens under the hood probably doesn't need to appear here.

Reply:
Equally, because we've added `BatchEncrypt` to `config.Encrypter` "all" `Encrypter`s are batching now, in some sense, but hey ho.

Reply:
Although only when constructed via a secrets manager .. and there's a whole bunch of places which call these methods directly with blinding crypters etc